### DISCUSSION TITLE
trick to use ttx with git to diff binary font files

### DISCUSSION THREAD

**Discussion by anthrotype** [fonttools/fonttools#2829] [General]
I just wanted to share a little trick I just learned which maybe others may find useful.
All credits go to this blog post, I just adapted it for ttx and font files: https://ten0s.github.io/blog/2020/04/19/git-difftool-and-binary-files

Basically there's a way to configure git to use a filter to create a texual representation of binary files and use that when doing `git diff` or `git show`, etc. Here I'll show how I set ttx to handle all files with .ttf and .otf extensions.

First I made a global `~/.gitattributes` file in my home directory with this content:

```
*.ttf diff=ttx
*.otf diff=ttx
```

then I edited my user's `~/.gitconfig` file by adding the following lines:

```
[core]
	# the section titled '[core]' may already be present in you .gitconfig, if so add the 'attributesfile' setting at the end of it;
	# this tells git where to find the user-global .gitattributes file that associates file extensions with out custom diff filter
	attributesfile = ~/.gitattributes

# this defines a filter named 'ttx', we can call it as we want, should only match the name used in the .gitattributes;
# The filter reads a binary file and dumps the content as text to standard output (-q to make ttx quiet, non verbose): 
[diff "ttx"]
	textconv = ttx -q -o -
```

Now, when I do `git diff` or `git show` or similar git commands, and there's a file with .ttf or .otf extensions, git automatically uses ttx before showing the textual diff. Pretty cool!

**Comment by anthrotype** [comment]
Now I wish we could make Github diff view also support this as a way to diff binary font files on the web. Does anybody know if that's possible?

**Comment by khaledhosny** [comment]
I’ve been doing this for a while. The only downside is that ttx is rather slow, so git log -p and co can take ages in repositories with big font files.

**Reply by anthrotype** [reply]
right.. i noticed that, after the first burst of enthusiasm 😅 
maybe One Day™ we shall have an oxidized version of ttx, who knows? 😉

**Reply by khaledhosny** [reply]
There is a tool called `ttfdump` that is ~10 times faster than ttx, but I think it is pretty old and does not support everything ttx supports, its output is also very low level and more like a text serialization of the binary.
